File: 02_double_dispatch/haskell/paper_scissor_rock.hs
data Element = Paper | Scissor | Rock 
  deriving (Show) 

beats :: Element -> Element -> Bool
beats Paper Scissor = False
beats Paper Rock    = True
beats Scissor Paper = True
beats Scissor Rock  = False
beats Rock Paper    = False
beats Rock Scissor  = True
beats _ _       = False

main :: IO ()
main = do
  print $ "paper beats paper? "    ++ show (Paper `beats` Paper)
  print $ "paper beats scissor? "  ++ show (Paper `beats` Scissor)
  print $ "paper beats rock? "     ++ show (Paper `beats` Rock)
  print $ "scissor beats paper? "  ++ show (Scissor `beats` Paper)
  print $ "scissor beats scissor? "++ show (Scissor `beats` Scissor)
  print $ "scissor beats rock? "   ++ show (Scissor `beats` Rock)
  print $ "rock beats paper? "     ++ show (Rock `beats` Paper)
  print $ "rock beats scissor? "   ++ show (Rock `beats` Scissor)
  print $ "rock beats rock? "      ++ show (Rock `beats` Rock)
  

File: 02_double_dispatch/scala/paper_scissor_rock.scala
sealed trait Element {
  def beats(other: Element): Boolean
}

case object Paper extends Element {
  def beats(other: Element): Boolean = other match {
    case Scissor => false
    case Rock => true
    case _ => false
  }
}

case object Scissor extends Element {
  def beats(other: Element): Boolean = other match {
    case Paper => true
    case Rock => false
    case _ => false
  }
}

case object Rock extends Element {
  def beats(other: Element): Boolean = other match {
    case Paper => false
    case Scissor => true
    case _ => false
  }
}

println("paper beats paper? "    + Paper.beats(Paper))
println("paper beats scissor? "  + Paper.beats(Scissor))
println("paper beats rock? "     + Paper.beats(Rock))
println("scissor beats paper? "  + Scissor.beats(Paper))
println("scissor beats scissor? "+ Scissor.beats(Scissor))
println("scissor beats rock? "   + Scissor.beats(Rock))
println("rock beats paper? "     + Rock.beats(Paper))
println("rock beats scissor? "   + Rock.beats(Scissor))
println("rock beats rock? "      + Rock.beats(Rock))
